### ISSUE TITLE
watchdog: multiple registrations of unmodified files

### ISSUE THREAD

**Issue by carueda** [ESIPFed/sweet-tools#3]
@lewismc notes the frequent email notifications about the registration of an ontology that hasn't been changed since the 3.3.0 release (28 days ago):

```
  A new ontology version has been registered:

     IRI: http://sweetontology.net/humanResearch
     Resolve with: http://cor.esipfed.org/ont/?iri=http://sweetontology.net/humanResearch
     Name: SWEET Ontology Human Research
     Version: 20190816T002947
     Owner: esip
     Submitter: sweetbot
     Updated: 2019-08-16T00:29:47.000Z
     Log: (not given)
```

**Comment by carueda** [comment]
Well, I've been running the watchdog directly on the COR box.

I first noted a timeout while completing the registration of an ontology. Upon increasing the timeout to 3min (from 1min), I'm now getting an error from Cloudflare:

```
[cor-admin1@ip-172-30-0-37 watchdog]$ ./watchdog.sc update
watchdog update starting Fri Aug 16 03:03:06 UTC 2019
Getting Github info: .................................................................................................................................................................................................................................
(225 paths retrieved)

3 changed ontologies:
	http://sweetontology.net/realmCryo
		- uploading
		- registering new revision of iri=http://sweetontology.net/realmCryo
error: registering new revision of iri=http://sweetontology.net/realmCryo: HttpResponse(<!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<meta http-equiv="refresh" content="0">

<title>cor.esipfed.org | 502: Bad gateway</title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" type="text/css" media="screen,projection" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" type="text/css" media="screen,projection" /><![endif]-->
<style type="text/css">body{margin:0;padding:0}</style>




</head>
<body>
<div id="cf-wrapper">



    <div id="cf-error-details" class="cf-error-details-wrapper">
        <div class="cf-wrapper cf-error-overview">
            <h1>

              <span class="cf-error-type">Error</span>
              <span class="cf-error-code">502</span>
              <small class="heading-ray-id">Ray ID: 50702c771e667971 &bull; 2019-08-16 03:04:11 UTC</small>
            </h1>
            <h2 class="cf-subheadline">Bad gateway</h2>
        </div><!-- /.error-overview -->

        <div class="cf-section cf-highlight cf-status-display">
            <div class="cf-wrapper">
                <div class="cf-columns cols-3">

<div id="cf-browser-status" class="cf-column cf-status-item cf-browser-status ">
  <div class="cf-icon-error-container">
    <i class="cf-icon cf-icon-browser"></i>
    <i class="cf-icon-status cf-icon-ok"></i>
  </div>
  <span class="cf-status-desc">You</span>
  <h3 class="cf-status-name">Browser</h3>
  <span class="cf-status-label">Working</span>
</div>

<div id="cf-cloudflare-status" class="cf-column cf-status-item cf-cloudflare-status ">
  <div class="cf-icon-error-container">
    <i class="cf-icon cf-icon-cloud"></i>
    <i class="cf-icon-status cf-icon-ok"></i>
  </div>
  <span class="cf-status-desc">Seattle</span>
  <h3 class="cf-status-name">Cloudflare</h3>
  <span class="cf-status-label">Working</span>
</div>

<div id="cf-host-status" class="cf-column cf-status-item cf-host-status cf-error-source">
  <div class="cf-icon-error-container">
    <i class="cf-icon cf-icon-server"></i>
    <i class="cf-icon-status cf-icon-error"></i>
  </div>
  <span class="cf-status-desc">cor.esipfed.org</span>
  <h3 class="cf-status-name">Host</h3>
  <span class="cf-status-label">Error</span>
</div>

                </div>

            </div>
        </div><!-- /.status-display -->

        <div class="cf-section cf-wrapper">
            <div class="cf-columns two">
                <div class="cf-column">
                    <h2>What happened?</h2>
                    <p>The web server reported a bad gateway error.</p>
                </div>

                <div class="cf-column">
                    <h2>What can I do?</h2>
                    <p>Please try again in a few minutes.</p>
                </div>
            </div>

        </div><!-- /.section -->

        <div class="cf-error-footer cf-wrapper">
  <p>
    <span class="cf-footer-item">Cloudflare Ray ID: <strong>50702c771e667971</strong></span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span>Your IP</span>: 34.216.150.176</span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span>Performance &amp; security by</span> <a href="https://www.cloudflare.com/5xx-error-landing?utm_source=error_footer" id="brand_link" target="_blank">Cloudflare</a></span>

  </p>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
</div><!-- /#cf-wrapper -->
</body>
</html>
,502,Map(Cache-Control -> Vector(no-store, no-cache, must-revalidate, post-check=0, pre-check=0), CF-RAY -> Vector(50702c771e667971-SEA), Connection -> Vector(keep-alive), Content-Type -> Vector(text/html; charset=UTF-8), Date -> Vector(Fri, 16 Aug 2019 03:04:11 GMT), Expires -> Vector(Thu, 01 Jan 1970 00:00:01 GMT), Pragma -> Vector(no-cache), Server -> Vector(cloudflare), Set-Cookie -> Vector(__cfduid=de00bc273ddc1ff84c4b0c444c0a264a21565924591; expires=Sat, 15-Aug-20 03:03:11 GMT; path=/; domain=.esipfed.org; HttpOnly, cf_ob_info=502:50702c771e667971:SEA; path=/; expires=Fri, 16-Aug-19 03:04:41 GMT, cf_use_ob=80; path=/; expires=Fri, 16-Aug-19 03:04:41 GMT), Status -> Vector(HTTP/1.1 502 Proxy Error), Transfer-Encoding -> Vector(chunked)))
```

Rendered, that html part is:

![image](https://user-images.githubusercontent.com/556505/63141118-5f394100-bf99-11e9-8671-d837707381e0.png)

This seems related with the error we also saw recently when trying to reload the triple store.

So, in summary so far, any operation involving the triple store (which includes ontology registration whose final step is to update the triple store once the ontology itself is stored in Mongo) seems to be triggering this error exposed by Cloudflare.  It might be that Cloudflare is rather picky regarding requests taking too long to respond, or something along those lines..

Now, what happens is that, because of this error, the watchdog program gets killed before it can update the local set of SHA-256 checksums, which are used for comparison purposes the next time it is run again.

**Comment by carueda** [comment]
@lewismc Those new emails just being sent are about repeated registrations, I know, but I'm testing my recent changes, that is, hoping that a subsequent run will not reattempt those same previous registrations.  Then I will see about unregistering those extra versions (at least to some of the files, you will be welcome to help with the cleanup ;)

**Comment by carueda** [comment]
Yes, these adjustments to the watchdog are working as intended. As added to the README:

```

    ./watchdog.sc update
    
NOTE: any error during a registration is just printed out but the process
is let to continue, in particular this means that the set of SHA-256
checksums will still be updated so the same registrations are not 
reattempted in a next run.  This logic is of course not very elegant/robust
but the case we want to handle at the moment is failures while updating 
the triple store, not error in the ontology registration itself.
```

These is the first run with a couple of the files not "refreshed" locally:

```
[cor-admin1@ip-172-30-0-37 watchdog]$ ./watchdog.sc update
Compiling /home/cor-admin1/sweet-tools/sc/watchdog/watchdog.sc
watchdog update starting Fri Aug 16 03:47:42 UTC 2019
Getting Github info: .................................................................................................................................................................................................................................
(225 paths retrieved)

3 changed ontologies:
	http://sweetontology.net/realmCryo
		- uploading
		- registering new revision of iri=http://sweetontology.net/realmCryo
error: registering new revision of iri=http://sweetontology.net/realmCryo: HttpResponse(<!DOCTYPE html>
...
NOTE: continuing...
	http://sweetontology.net/reprMathStatistics
		- uploading
		- registering new revision of iri=http://sweetontology.net/reprMathStatistics
error: registering new revision of iri=http://sweetontology.net/reprMathStatistics: HttpResponse(<!DOCTYPE html>
...
NOTE: continuing...
	http://sweetontology.net/reprSciModel
		- uploading
		- registering new revision of iri=http://sweetontology.net/reprSciModel
error: registering new revision of iri=http://sweetontology.net/reprSciModel: HttpResponse(<!DOCTYPE html>
...
NOTE: continuing...

1 removed ontologies:
	http://sweetontology.net/alignment/dcat
		unregistration not implemented.
Updating local info... Done.
```

Then, an immediate 2nd run:

```
[cor-admin1@ip-172-30-0-37 watchdog]$ ./watchdog.sc update
watchdog update starting Fri Aug 16 03:53:51 UTC 2019
Getting Github info: .................................................................................................................................................................................................................................
(225 paths retrieved)

1 removed ontologies:
	http://sweetontology.net/alignment/dcat
		unregistration not implemented.
Updating local info... Done.
```

**Comment by carueda** [comment]
This should be resolved already, in particular because of various adjustments done while addressing the cloudflare restrictions issue, but also because cloudflare itself was removed on the COR box.
